src/requests.rs

```diff
@@ -1,18 +1,33 @@
-use crate::primitives::{BodyRequest, DecodeError, FromByte};
+use crate::primitives::{
+    decode_buffer, ApiVersion, BodyRequest, DecodeError, FromByte, FromByteWithVersion,
+};
 use bytes::Buf;
 
 #[derive(PartialEq, Debug)]
 pub struct BodyApiVersionRequest {}
-impl FromByte for BodyApiVersionRequest {
-    fn decode(_: &mut Buf) -> Result<Self, DecodeError> {
+impl FromByteWithVersion for BodyApiVersionRequest {
+    fn decode_with_version(_: &mut Buf, _: i16) -> Result<Self, DecodeError> {
         Ok(BodyApiVersionRequest {})
     }
 }
 
 #[derive(PartialEq, Debug)]
-pub struct BodyMetadataRequest {}
-impl FromByte for BodyMetadataRequest {
-    fn decode(_: &mut Buf) -> Result<Self, DecodeError> {
-        Ok(BodyMetadataRequest {})
+pub struct BodyMetadataRequest {
+    pub topics: String,
+    pub allow_auto_topic_creation: bool,
+}
+impl FromByteWithVersion for BodyMetadataRequest {
+    fn decode_with_version(buf: &mut Buf, api_version: i16) -> Result<Self, DecodeError> {
+        if api_version <= 3 {
+            Ok(BodyMetadataRequest {
+                topics: decode_buffer(buf).unwrap_or(String::from("")),
```

Comment:
If a valid string couldn't be read from the buffer, this should probably fail, rather than using an empty string?
```suggestion
                topics: decode_buffer(buf)?,
```

Reply:
The `topics` can be **optional** https://kafka.apache.org/protocol.html#The_Messages_Metadata . I'll check how the official Kafka client handles this. This is weird to me because in most cases the optional fields should be at the end of the buffer per my understanding.

Reply:
It's pretty interesting to see the comments in the official client codebase about the empty string vs null here:

```
    // In version 0, an empty array indicates "request metadata for all topics."  In version 1 and
    // higher, an empty array indicates "request metadata for no topics," and a null array is used to
    // indiate "request metadata for all topics."
```

https://github.com/apache/kafka/blob/5144660040839cee6e213b5146c1ae44340eddb0/clients/src/main/resources/common/message/MetadataRequest.json#L22-L24

Reply:
Oh! I just realized that the `[]` means `KAFKA Array` **rather than optional**...

Reply:
Good news: the `serializedBytes` of the requests in the official Kafka clients is at https://github.com/apache/kafka/blob/6dd4ebcea78f9c6fb8a2921cab3fd92be4978063/core/src/test/scala/unit/kafka/server/BaseRequestTest.scala#L133 .

We can run the test cases in the official Kafka client codebase and test our own codec against the serialized bytes.